File: test/problems/3-hard/239-Sliding-Window-Maximum/solution.cpp
#include <vector>
#include <deque>
using namespace std;

vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    deque<int> max_dq;
    vector<int> max_sliding_window;
    int l = 0;
    int r = 0;
    while (r < nums.size()) {
        // shrink left by if length == k
        if ( (r-l) == k ) {
            if (max_dq.front() == nums[l]) {
                max_dq.pop_front();
            }
            l += 1;
        }

        // increase right by 1
        while (!max_dq.empty() && max_dq.back() < nums[r]) {
            max_dq.pop_back();
        }
        max_dq.push_back(nums[r]);
        r += 1;

        // append to answer list if length == k
        if ( (r-l) == k) {
            max_sliding_window.push_back(max_dq.front());
        }

    }
    
    return max_sliding_window;
}

File: test/problems/3-hard/239-Sliding-Window-Maximum/solution.java
import java.util.*;

public int[] maxSlidingWindow(int[] nums, int k) {
    Deque<Integer> maxDq = new ArrayDeque<>();
    List<Integer> maxSlidingWindow = new ArrayList<>();
    int l = 0;
    int r = 0;
    
    while (r < nums.length) {
        // Shrink left if length == k
        if ((r - l) == k) {
            if (!maxDq.isEmpty() && maxDq.peekFirst() == nums[l]) {
                maxDq.pollFirst();
            }
            l++;
        }
        
        // Increase right by 1
        while (!maxDq.isEmpty() && maxDq.peekLast() < nums[r]) {
            maxDq.pollLast();
        }
        maxDq.offerLast(nums[r]);
        r++;
        
        // Append to answer list if length == k
        if ((r - l) == k) {
            maxSlidingWindow.add(maxDq.peekFirst());
        }
    }
    
    return maxSlidingWindow.stream().mapToInt(i -> i).toArray();
}